### PR TITLE
feat: add devcontainer

### PR DESCRIPTION
**Pull request by domoritz**

### CODE REVIEW

**Comment by domoritz** [comment]
Alternative to https://github.com/vega/vega-lite/pull/7807/files

File: .devcontainer/entrypoint.sh [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env bash
```
Comment (behrica) [inline .devcontainer/entrypoint.sh:1]: This seems against good practices in Docker.
To  use an "entrypoint" for the purpose of validating the Dockerfile, even if conditional.
I would do that with  a github action:
```
docker-check:
      name: docker
      runs-on: ubuntu-latest
      steps:
      - uses: actions/checkout@v2
        with:
          token: ${{ secrets.GH_PAT || github.token }}

      - name: Build Dockerfile
        run: docker build -t vega-lite .
        
      - name: Run yarn-build
        run: docker run -ti -v `pwd`:/app -w /app vega-lite yarn build
```

Comment (domoritz) [reply .devcontainer/entrypoint.sh]: Good point. I took a different approach and built the docker file as you suggested.

**Comment by domoritz** [comment]
@behrica can you test `yarn site` and `yarn test:runtime` in the container. I can't get the latter to work yet.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -134,6 +134,11 @@ For Jekyll and its dependencies, because we already have the `Gemfile` in the re
 pushd site && bundle install && popd
 ```
 
+### Develop inside a Container
+
+We support (optional) development of Vega-Lite in [VSCode Remote containers](https://code.visualstudio.com/docs/remote/containers). Once you have the [Remote Development extension pack](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.vscode-remote-extensionpack) installed,
+opening the Vega-Lite source folder, will propose you to re-open it in a container. Once the image is built, you can tun the development commands inside the container.
+
```
Comment (domoritz) [inline CONTRIBUTING.md:141]: Thank you @behrica

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -134,6 +134,11 @@ For Jekyll and its dependencies, because we already have the `Gemfile` in the re
 pushd site && bundle install && popd
 ```
 
+### Develop inside a Container
+
+We support (optional) development of Vega-Lite in [VSCode Remote containers](https://code.visualstudio.com/docs/remote/containers). Once you have the [Remote Development extension pack](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.vscode-remote-extensionpack) installed,
+opening the Vega-Lite source folder, will propose you to re-open it in a container. Once the image is built, you can tun the development commands inside the container.
```
Comment (behrica) [inline CONTRIBUTING.md]: "tun" -> "tune"

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -136,8 +136,7 @@ pushd site && bundle install && popd
 
 ### Develop inside a Container
 
-We support (optional) development of Vega-Lite in [VSCode Remote containers](https://code.visualstudio.com/docs/remote/containers). Once you have the [Remote Development extension pack](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.vscode-remote-extensionpack) installed,
-opening the Vega-Lite source folder, will propose you to re-open it in a container. Once the image is built, you can tun the development commands inside the container.
+We support (optional) development of Vega-Lite in [VSCode Remote containers](https://code.visualstudio.com/docs/remote/containers). Once you have the [Remote Development extension pack](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.vscode-remote-extensionpack) installed, opening the Vega-Lite source folder, will propose you to re-open it in a container. Once the image is built, you can tun the development commands inside the container.
```
Comment (behrica) [inline CONTRIBUTING.md:139]: "tun" -> "tune"

File: CONTRIBUTING.md (hunk 3 above)
Comment (domoritz) [reply CONTRIBUTING.md]: It was supposed to be `run`

**Comment by domoritz** [comment]
@behrica I get `Error: Jest: Got error running globalSetup - /workspaces/vega-lite/test-runtime/setup.js, reason: Failed to launch the browser process! spawn /usr/bin/chromium-browser ENOENT` when I try to run `yarn test:runtime`. Any idea how to fix it?

**Comment by behrica** [comment]
For me they fail differently, all svg's are different.
Can you try to repeat from scratch
1. cd /tmp && git clone https://github.com/vega/vega-lite.git && cd vega-lite && git checkout origin/dom/devcontainer
2. code .
3.  -> "Reopen in Container"
4. wait ....
5. "open terminal" 
5. yarn build && yarn test:runtime

This gives me:

test Suites: 6 failed, 6 total
Tests:       50 failed, 50 total
Snapshots:   0 total
Time:        3.433 s
All fail with "unexpected svg" (which looks "valid" svg)

https://gist.github.com/behrica/4cecd24b39988f42baea67c1f6047f0c

**Comment by behrica** [comment]
The error you had (or similar at least) was fixed by:
https://github.com/vega/vega-lite/blob/63233b2d66993503a64f4b9c02b839e7d408fd37/test-runtime/setup.js#L21

**Comment by behrica** [comment]
I am pretty sure, that I had the tests once working in my PR.
Will go back to it and check.

**Comment by behrica** [comment]
My failing tests above have tiny differences in the svg:

```
90c190
< transform=\"translate(-61.0869140625,100)\"><path 
---
> transform=\"translate(-58.6845703125,100)\"><path 
263c263
< transform=\"translate(-30.0869140625,100) 
---
> transform=\"translate(-27.6845703125,100) 
272c272
< transform=\"translate(-61.0869140625,100)\"><path 
---
> transform=\"translate(-58.6845703125,100)\"><path 
693c693
< d=\"M0,0h23v53h-23Z\" 
---
> d=\"M0,0h22v53h-22Z\" 
704c704
< d=\"M0,0h22.3623046875v11h-22.3623046875Z\" 
---
> d=\"M0,0h21.5615234375v11h-21.5615234375Z\" 
721c721
< d=\"M0,0h22.3623046875v11h-22.3623046875Z\" 
---
> d=\"M0,0h21.5615234375v11h-21.5615234375Z\" 
738c738
< d=\"M0,0h22.3623046875v11h-22.362304687
```

**Comment by domoritz** [comment]
Hmm, that shouldn't happen since we run inside puppeteer. Maybe the fonts aren't working inside the container?

**Comment by behrica** [comment]
They pass on my branch. That is strange as well.
The "commands" run are the same.
Slight differences are: 
- ruby 2.5 vs 2.7
- order of yarn commands
- which users runs them is eventually different ('node' vs 'root') not sure.

**Comment by behrica** [comment]
Maybe reason for difference:

node 14 vs node 16

**Comment by behrica** [comment]
> Hmm, that shouldn't happen since we run inside puppeteer. Maybe the fonts aren't working inside the container?

You mean that certain "fonts are eventually not installed inside container " ? Could be.

**Comment by domoritz** [comment]
Yep.

**Comment by behrica** [comment]
I try.

**Comment by behrica** [comment]
Which "fonts" could be needed ? Did not find any docu on this ?

File: .devcontainer/devcontainer.json [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+// https://github.com/microsoft/vscode-dev-containers/tree/v0.205.2/containers/typescript-node
+{
+  "name": "Node.js & TypeScript",
+  "build": {
+    "dockerfile": "Dockerfile",
+    // Update 'VARIANT' to pick a Node version: 16, 14, 12.
+    // Append -bullseye or -buster to pin to an OS version.
+    // Use -bullseye variants on local on arm64/Apple Silicon.
+    "args": {
+      "VARIANT": "16-bullseye"
```
Comment (behrica) [inline .devcontainer/devcontainer.json:11]: This setting get overwritten in Dockerfile. One of the 2 should be removed.

File: .devcontainer/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,18 @@
+# See here for image contents: https://github.com/microsoft/vscode-dev-containers/tree/v0.205.2/containers/typescript-node/.devcontainer/base.Dockerfile
+
+# [Choice] Node.js version (use -bullseye variants on local arm64/Apple Silicon): 16, 14, 12, 16-bullseye, 14-bullseye, 12-bullseye, 16-buster, 14-buster, 12-buster
+ARG VARIANT="16-bullseye"
```
Comment (behrica) [inline .devcontainer/Dockerfile:4]: should be: "ARG VARIANT"   , will then be set from devcontainer.json

File: .devcontainer/devcontainer.json (hunk 5 above)
Comment (domoritz) [reply .devcontainer/devcontainer.json:11]: I just used what vscode auto generated. Which one should be replaced? Does it fix the issues?

File: .devcontainer/Dockerfile (hunk 6 above)
Comment (domoritz) [reply .devcontainer/Dockerfile:4]: I'm not following

Comment (behrica) [reply .devcontainer/Dockerfile:4]: Sorry, I read again the docker docu. It is correct. It will be overwritten by whatever is in devcontainer.json

**Comment by behrica** [comment]
I cannot figure it out, why my Dockerfile works. It did variations of anything imaginle to yours (taking from mine), but is still fails.
It only works from outside "code", with a commandline:

```
docker run -ti  -u 1000:1000 -v `pwd`:/vega-lite -w /vega-lite cc055a57b3c4 bash -c "pushd site && bundle install && popd && yarn test:runtime"
```

Where docker image cc055a57b3c4 is just a `docker build` of  https://github.com/behrica/vega-lite/blob/ae20ddcbcfbf6d822aa689d4b01816e7b999fc1f/Dockerfile

**Comment by behrica** [comment]
Could you maybe try to run this, just to double check:

```
docker build -t vega-lite-add-docker https://raw.githubusercontent.com/behrica/vega-lite/ae20ddcbcfbf6d822aa689d4b01816e7b999fc1f/Dockerfile
docker run -ti  -u 1000:1000 -v `pwd`:/vega-lite -w /vega-lite vega-lite-add-docker /bin/bash -c "pushd site && bundle install && popd && yarn test:runtime"
```

**Comment by domoritz** [comment]
I get this in fish shell

```
docker: Error response from daemon: create `pwd`: "`pwd`" includes invalid characters for a local volume name, only "[a-zA-Z0-9][a-zA-Z0-9_.-]" are allowed. If you intended to pass a host directory, use absolute path.
See 'docker run --help'.
278:284: syntax error: A identifier can’t go after this “"”. (-2740)
```

and 

```
docker run -ti  -u 1000:1000 -v 'pwd':/vega-lite -w /vega-lite vega-lite-add-docker /bin/bash -c "pushd site && bundle install && popd && yarn test:runtime"
/bin/bash: line 0: pushd: site: No such file or directory
```

if I fix the quotes.

**Comment by behrica** [comment]
I think by chance I was "lucky" to get a working setup.
But I have no idea any more, how to change this PR to make the tests work.

I propose we continue with my PR, adding changes if needed, while making sure, that build still passes.
Mine runs as well "yarn test:runtime" in VSCode without errors

**Comment by domoritz** [comment]
Okay, can you integrate reasonable changes from my pull request? For example, I shortened the instructions in the contributing guide, used recommended extensions, and "features" in the devcontainer file.

**Comment by domoritz** [comment]
Okay. We can use your setup instead. I sent a review.

File: .devcontainer/devcontainer.json (hunk 5 above)
Comment (behrica) [reply .devcontainer/devcontainer.json:11]: its fine, I understood wrong

**Pull request closed without merging**